Surface builder thinking content in thinking stall error message for diagnostics

## Problem

When a builder thinking stall is detected, the shepherd error message contains no information about what the builder was reasoning about. The builder log is filled with thousands of raw \`thinking\` tokens, providing zero diagnostic value for understanding why the stall occurred or whether the issue itself is the root cause.

## Observed Behavior

Builder log for a thinking stall contains only:
\`\`\`
thinking
thinking
thinking
Flambéing…
thinking
thinking
...
\`\`\`

Shepherd error message:
\`\`\`
[ERROR] builder thinking stall: extended thinking output with zero tool calls detected (not retryable)
\`\`\`

No information about what the builder was thinking, making it impossible to determine:
- Was the issue spec too ambiguous?
- Was the builder confused about the codebase?
- Was this a model-level transient failure?

## Expected Behavior

When a thinking stall is detected, the shepherd should extract and surface a summary of the builder's thinking content in:
1. The shepherd error log message
2. A post-mortem comment or structured output that can guide issue triage

## Proposed Fix

Capture the actual thinking text content (not just the token stream) from the builder session. When a stall is detected, include the last N characters of thinking content in the error message. This would allow humans and the loom daemon to make better decisions about whether to retry, modify the issue, or escalate.

## References

- Observed during shepherd runs for #2812
- Builder log: \`.loom/logs/loom-builder-issue-2812.log\`

--- a/loom-tools/src/loom_tools/shepherd/phases/base.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/base.py
@@ -1185,6 +1185,46 @@ def _scan_log_for_thinking_stall(log_path: Path) -> bool:
         return False
 
 
+# Maximum number of characters extracted from a thinking stall log for
+# inclusion in diagnostic messages.  Large enough to be useful, small enough
+# to avoid flooding the shepherd log.
+_THINKING_SNIPPET_MAX_CHARS = 500
+
+
+def _extract_thinking_snippet(log_path: Path, max_chars: int = _THINKING_SNIPPET_MAX_CHARS) -> str:
+    """Extract a diagnostic snippet from a thinking-stall builder log.
+
+    Reads the builder log, strips ANSI codes, and returns the tail of
+    the CLI output (up to *max_chars* characters).  Even though the
+    log typically contains only spinner animations during a thinking stall,
+    the tail gives shepherds and human operators a concrete view of the
+    session state and distinguishes true thinking stalls from early crashes.
+
+    Args:
+        log_path: Path to the builder session log file.
+        max_chars: Maximum number of characters to return.
+
+    Returns:
+        A stripped string of the last ``max_chars`` characters of CLI output,
+        or an empty string if the log is unavailable or empty.
+    """
+    if not log_path.is_file():
+        return ""
+
+    try:
+        content = log_path.read_text()
+        stripped = strip_ansi(content)
+        cli_output = _get_cli_output(stripped)
+        if not cli_output:
+            return ""
+
+        tail = cli_output.strip()[-max_chars:]
+        return tail.strip()
+
+    except OSError:
+        return ""
+
+
 def _classify_low_output_cause(log_path: Path) -> str:
     """Classify the root cause of a low-output session from the worker log.
 
@@ -1744,10 +1784,16 @@ def run_worker_phase(
             thinking_log_path = _session_log_path(ctx.repo_root, actual_name)
             if _scan_log_for_thinking_stall(thinking_log_path):
                 elapsed_s = int(time.monotonic() - phase_start)
+                thinking_snippet = _extract_thinking_snippet(thinking_log_path)
+                snippet_suffix = (
+                    f"\nBuilder thinking tail:\n{thinking_snippet}"
+                    if thinking_snippet
+                    else ""
+                )
                 log_warning(
                     f"Thinking stall detected: {role} session '{actual_name}' "
                     f"has produced output for {elapsed_s}s with zero tool calls, "
-                    f"terminating early (log: {thinking_log_path})"
+                    f"terminating early (log: {thinking_log_path}){snippet_suffix}"
                 )
                 wait_proc.terminate()
                 wait_proc.wait(timeout=30)

--- a/loom-tools/src/loom_tools/shepherd/phases/builder.py
+++ b/loom-tools/src/loom_tools/shepherd/phases/builder.py
@@ -45,6 +45,7 @@ from loom_tools.shepherd.phases.base import (
     PhaseResult,
     PhaseStatus,
     _STARTUP_MONITOR_MCP_RESOLUTION,
+    _extract_thinking_snippet,
     _get_cli_output,
     extract_log_errors,
     run_phase_with_retry,
@@ -547,9 +548,17 @@ class BuilderPhase:
             # Thinking stall: builder produced output (spinner/thinking)
             # but never made a tool call.  Not retryable — the same
             # conditions will produce the same result.  See issue #2784.
+            log_path = self._get_log_path(ctx)
+            thinking_snippet = _extract_thinking_snippet(log_path)
+            snippet_suffix = (
+                f"\nBuilder thinking tail:\n{thinking_snippet}"
+                if thinking_snippet
+                else ""
+            )
             log_warning(
                 f"Thinking stall for issue #{ctx.config.issue}: "
                 f"builder produced thinking output but zero tool calls"
+                f"{snippet_suffix}"
             )
             self._cleanup_stale_worktree(ctx)
             return PhaseResult(
@@ -562,7 +571,8 @@ class BuilderPhase:
                 data={
                     "thinking_stall": True,
                     "exit_code": exit_code,
-                    "log_file": str(self._get_log_path(ctx)),
+                    "log_file": str(log_path),
+                    "thinking_snippet": thinking_snippet,
                 },
             )
 

--- a/loom-tools/tests/shepherd/test_phases.py
+++ b/loom-tools/tests/shepherd/test_phases.py
@@ -42,6 +42,7 @@ from loom_tools.shepherd.phases.base import (
     _RATE_LIMIT_SENTINEL,
     _classify_ghost_cause,
     _classify_low_output_cause,
+    _extract_thinking_snippet,
     _is_auth_failure,
     _is_ghost_session,
     _is_rate_limit_abort,
@@ -1055,6 +1056,118 @@ class TestBuilderPhase:
         assert "stall" in result.message.lower()
         assert result.data.get("planning_stall") is True
         assert result.data.get("planning_timeout") == 600
+
+    def test_thinking_stall_returns_failed_with_snippet(
+        self, mock_context: MagicMock, tmp_path: Path
+    ) -> None:
+        """Builder exit code 14 should return FAILED with thinking_snippet in data.
+
+        When the builder session hits a thinking stall (extended thinking with
+        zero tool calls), the shepherd should capture the tail of the builder
+        log and include it in the PhaseResult data for diagnostic purposes.
+        See issue #2855.
+        """
+        mock_context.check_shutdown.return_value = False
+        mock_context.config.planning_timeout = 600
+        mock_context.repo_root = tmp_path
+        wt_mock = MagicMock()
+        wt_mock.is_dir.return_value = True
+        wt_mock.__bool__ = lambda self: True
+        mock_context.worktree_path = wt_mock
+
+        # Create a builder log with simulated thinking output
+        log_dir = tmp_path / ".loom" / "logs"
+        log_dir.mkdir(parents=True)
+        log_file = log_dir / "loom-builder-issue-42.log"
+        thinking_content = (
+            "# CLAUDE_CLI_START\n"
+            + "Embellishing…\n" * 30
+            + "Moseying…\n" * 20
+        )
+        log_file.write_text(thinking_content)
+
+        builder = BuilderPhase()
+
+        with (
+            patch(
+                "loom_tools.shepherd.phases.builder.get_pr_for_issue", return_value=None
+            ),
+            patch("loom_tools.shepherd.phases.builder.transition_issue_labels"),
+            patch(
+                "loom_tools.shepherd.phases.builder.run_phase_with_retry",
+                return_value=14,
+            ),
+            patch.object(builder, "_cleanup_stale_worktree"),
+            patch.object(builder, "_create_worktree_marker"),
+        ):
+            result = builder.run(mock_context)
+
+        assert result.status == PhaseStatus.FAILED
+        assert "thinking stall" in result.message
+        assert result.data.get("thinking_stall") is True
+        assert "thinking_snippet" in result.data
+        # Snippet should contain the tail of the thinking output
+        snippet = result.data["thinking_snippet"]
+        assert isinstance(snippet, str)
+        assert len(snippet) > 0
+        assert "Moseying" in snippet
+
+
+class TestExtractThinkingSnippet:
+    """Unit tests for _extract_thinking_snippet helper."""
+
+    def test_returns_empty_for_missing_file(self, tmp_path: Path) -> None:
+        """Should return empty string when log file does not exist."""
+        result = _extract_thinking_snippet(tmp_path / "nonexistent.log")
+        assert result == ""
+
+    def test_returns_empty_for_no_sentinel(self, tmp_path: Path) -> None:
+        """Should return empty string when log has no CLI start sentinel."""
+        log_file = tmp_path / "test.log"
+        log_file.write_text("some output\nbut no sentinel\n")
+        result = _extract_thinking_snippet(log_file)
+        assert result == ""
+
+    def test_extracts_tail_from_thinking_output(self, tmp_path: Path) -> None:
+        """Should extract tail of CLI output after the sentinel."""
+        log_file = tmp_path / "test.log"
+        log_file.write_text(
+            "# log header\n"
+            "# CLAUDE_CLI_START\n"
+            + "Embellishing…\n" * 50
+            + "Moseying…\n" * 10
+        )
+        result = _extract_thinking_snippet(log_file, max_chars=200)
+        assert isinstance(result, str)
+        assert len(result) <= 200
+        assert "Moseying" in result
+
+    def test_respects_max_chars(self, tmp_path: Path) -> None:
+        """Should not return more than max_chars characters."""
+        log_file = tmp_path / "test.log"
+        long_content = "# CLAUDE_CLI_START\n" + "X" * 2000 + "\n"
+        log_file.write_text(long_content)
+        result = _extract_thinking_snippet(log_file, max_chars=100)
+        assert len(result) <= 100
+
+    def test_strips_ansi_codes(self, tmp_path: Path) -> None:
+        """Should strip ANSI escape codes from log content."""
+        log_file = tmp_path / "test.log"
+        # Write content with ANSI colour codes
+        log_file.write_text(
+            "# CLAUDE_CLI_START\n"
+            "\x1b[32mEmbellishing…\x1b[0m\n" * 10
+        )
+        result = _extract_thinking_snippet(log_file)
+        assert "\x1b" not in result
+        assert "Embellishing" in result
+
+    def test_returns_empty_for_empty_cli_output(self, tmp_path: Path) -> None:
+        """Should return empty string when there is no CLI output after sentinel."""
+        log_file = tmp_path / "test.log"
+        log_file.write_text("# CLAUDE_CLI_START\n# only header lines\n")
+        result = _extract_thinking_snippet(log_file)
+        assert result == ""
 
 
 class TestBuilderDiagnostics:

